oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AccessTokenResponseClient.java

```diff
@@ -0,0 +1,290 @@
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
+import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.ResponseErrorHandler;
+import org.springframework.web.client.RestOperations;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE;
+
+/**
+ * Base implementation of an {@link OAuth2AccessTokenResponseClient}
+ * that "exchanges" an authorization grant credential for an access token credential
+ * at the Authorization Server's Token Endpoint.
+ * This implementation uses a {@link RestOperations} when requesting the Access Token Response.
+ *
+ * @author Joe Grandja
+ * @since 5.1
+ * @see OAuth2AccessTokenResponseClient
+ * @see AbstractOAuth2AuthorizationGrantRequest
+ * @see OAuth2AccessTokenResponse
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-5.1">Section 5.1 Access Token Response</a>
+ */
+public abstract class AbstractOAuth2AccessTokenResponseClient<T extends AbstractOAuth2AuthorizationGrantRequest> implements OAuth2AccessTokenResponseClient<T> {
+	protected static final String INVALID_TOKEN_REQUEST_ERROR_CODE = "invalid_token_request";
+
+	protected static final String INVALID_TOKEN_RESPONSE_ERROR_CODE = "invalid_token_response";
+
+	protected static final String[] TOKEN_RESPONSE_PARAMETER_NAMES = {
+			OAuth2ParameterNames.ACCESS_TOKEN,
+			OAuth2ParameterNames.TOKEN_TYPE,
+			OAuth2ParameterNames.EXPIRES_IN,
+			OAuth2ParameterNames.REFRESH_TOKEN,
+			OAuth2ParameterNames.SCOPE
+	};
+
+	protected RestOperations restOperations;
+
+	protected AbstractOAuth2AccessTokenResponseClient() {
+		RestTemplate restTemplate = new RestTemplate();
+		// Disable the ResponseErrorHandler as errors are handled directly within this class
+		restTemplate.setErrorHandler(new NoOpResponseErrorHandler());
+		this.restOperations = restTemplate;
+	}
+
+	@Override
+	public OAuth2AccessTokenResponse getTokenResponse(T authorizationGrantRequest) throws OAuth2AuthenticationException {
```

Comment:
Rename parameter `authorizationGrantRequest` to `accessTokenRequest` and mark it `final`